### ISSUE TITLE
[Feature Request] Output rules used during parsing

### ISSUE THREAD

**Issue by d4l3k** [mna/pigeon#72]
Sometimes with really large grammars (such as the wikitext/parsoid peg definitions) it's really hard to debug which rules are actually being used to output the final result. It'd be really nice if pigeon could return a tree of which rules are being used to generate the final result.

**Comment by breml** [comment]
@d4l3k I not yet understand, what particular functionality your are looking for. How is the tree you are requesting different (or related) to the output you get, if the `Debug` option is set for the parser?
Is your request only about the actions where Go code is executed? In your `debug.go` you only modify `actionExpr`, but you seem to ignore to other code expressions (like `AndCodeExpr`, `NotCodeExpr`, `StateCodeExpr`).
Are you looking for a way to overwrite (or provide your own version of) the `callFuncTemplates` (see https://github.com/mna/pigeon/blob/master/builder/builder.go#L35)?

**Comment by d4l3k** [comment]
Yep, I was referring to printing the matches. There's another go PEG library which has a PrintSyntaxTree method which just prints each AST node's name and start/end.

https://github.com/pointlander/peg/blob/master/peg.peg.go#L292-L298

**Comment by breml** [comment]
OK, I understand.

What would be needed for this?

* Introduce a new option for the parser to instruct it to output the matched rules / syntax tree, maybe allow to provide some configuration, e.g. what should be printed (rule name, start-end index, matched input / runes, name of the called Go function, etc.)
* Is output of this to stdout enough or do we need to provide other options, like accepting a `Writer` interface to allow the user of the parser redirect the output to stdout/file/buffer/etc.
* Define an output format for this: should it be primarely human-readable or are there other use cases, where it would be interesting to process this output also with tools (grep/awk) or even JSON).
* Should the output format be configurable (template?)
* Is it only about the action expressions of the matched rules or is it also about the other expressions?

From the example @d4l3k provided I understand that the prettyprint function in the end prints a (colored) line per "node" (matched rule) in the grammar (see https://github.com/pointlander/peg/blob/master/peg.peg.go#L234).

In https://github.com/d4l3k/wikigopher/blob/master/wikitext/debug.go only the name of the called Go function is printed.

I feel like we first have to outline, what exactly should be implemented. Then we can work towards a PR. @d4l3k would you be willing to work on such a PR?

**Comment by d4l3k** [comment]
I think adding in a similar AST struct output would make sense since you
have can't print it as you visit anyways. This supports custom printing as
well.

My implemention is much less than ideal, just was the easiest way to get
something workable. I'd be interested in putting together a PR.

```go
type AST {
  Rule *rule
  Start, End int
  Match []byte
  Children []*AST
}

func ParseAST(...)
```

A separate parse AST function could also be useful since some times a user
may only want the PEG ast instead of the parser output
